source/common/config/metadata.h

```diff
@@ -46,5 +50,28 @@ class Metadata {
                                                   const std::string& key);
 };
 
+template <typename factoryClass> class TypedMetadataImpl : public TypedMetadata {
+public:
+  static_assert(std::is_base_of<Config::TypedMetadataFactory, factoryClass>::value,
+                "Factory type must be inherited from Envoy::Config::TypedMetadataFactory.");
+  TypedMetadataImpl(const ::envoy::api::v2::core::Metadata& metadata) : data_() {
+    auto& dataByKey = metadata.filter_metadata();
+    for (const auto& it : Registry::FactoryRegistry<factoryClass>::factories()) {
+      const auto& metaIter = dataByKey.find(it.first);
+      if (metaIter != dataByKey.end()) {
+        data_[it.second->name()] = it.second->parse(metaIter->second);
+      }
+    }
+  }
+
+  const StreamInfo::FilterState::Object* getData(const std::string& key) const override {
+    const auto& it = data_.find(key);
+    return it == data_.end() ? nullptr : it->second.get();
+  }
+
+private:
+  std::map<std::string, std::unique_ptr<const StreamInfo::FilterState::Object>> data_;
```

Comment:
Does this map need to be ordered?

Reply:
no, change to unordered_map? Note unordered_map worst case is O(N) though.